### PR TITLE
Add SpecificityAtSensitivity Metric

### LINKED ISSUE #971: Add classification metrics: `SensitivityAtSpecificity` and `SpecificityAtSensitivity`
### Discussed in https://github.com/PyTorchLightning/metrics/discussions/965

<div type='discussions-op-text'>

<sup>Originally posted by **iamharsha1999** April 19, 2022</sup>
I was specifically looking for the following two metrics:
1. Sensitivity at Specificity
2. Specificity at Sensitivity

However I was not able to find it ? Would it be possible to add these implementations to torchmetrics. 

You can find the tensorflow implementations below:
https://www.tensorflow.org/api_docs/python/tf/keras/metrics/SpecificityAtSensitivity
https://www.tensorflow.org/api_docs/python/tf/keras/metrics/SensitivityAtSpecificity

### PR DESCRIPTION
**Pull request by shenoynikhil**
## What does this PR do?

Partially Fixes #971. Introduces the `SpecificityAtSensitivity` Metric. The computation is as follows,
1. Compute false positive rate, true positive rate at different thresholds (Using the roc curve)
2. true positive rate is sensitivity while `1 - false positive rate` is specificity
3. Consider all thresholds where sensitivity >= min_sensitivity (this is provided by the user)
4. Calculate the maximum possible specificity at all the thresholds

#### PseudoCode
```python
from torchmetrics.functional.classification.roc import _binary_roc_compute

preds, target, thresholds = _binary_precision_recall_curve_format(preds, target, thresholds, ignore_index)
state = _binary_precision_recall_curve_update(preds, target, thresholds)
fpr, sensitivity, thresholds = _binary_roc_compute(state, thresholds, pos_label)
specificity = 1 - fpr

# actual computation of max specificity at min possible sensitivity
max_spec, _, best_threshold = max(
    [(sp, sn, thresh) for sp, sn, thresh in zip(specificity, sensitivity, thresholds) if sn >= min_sensitivity]
)

return max_spec, best_threshold
```

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by shenoynikhil** [comment]
@Borda @stancld @SkafteNicki If you can provide some feedback on the edge cases (all positive and all negative), that would be great. I just wrote the edge case based on the `recall_at_fixed_precision metric`.

File: src/torchmetrics/functional/classification/specificity_at_sensitivity.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+
+
+def _specificity_at_sensitivity(
+    specificity: Tensor,
+    sensitivity: Tensor,
+    thresholds: Tensor,
+    min_sensitivity: float,
+) -> Tuple[Tensor, Tensor]:
+    try:
+        max_spec, _, best_threshold = max(
```
Comment (justusschock) [inline src/torchmetrics/functional/classification/specificity_at_sensitivity.py:53]: can we make this more explicit that it uses the first item of the tuple for comparison? If they're equal it would use the second item which does not make sense, right?

**Review by justusschock — APPROVED** [review]
Edge-Cases look good to me. @Borda errors are unrelated. Will investigate.

File: src/torchmetrics/functional/classification/specificity_at_sensitivity.py (hunk 1 above)
Comment (shenoynikhil) [reply src/torchmetrics/functional/classification/specificity_at_sensitivity.py:53]: That makes sense. Let me have a look at it.

Comment (shenoynikhil) [reply src/torchmetrics/functional/classification/specificity_at_sensitivity.py:53]: The list which we call the max on is this,
```python
[
   (sp, sn, thresh) for sp, sn, thresh in zip(specificity, sensitivity, thresholds) 
   if sn >= min_sensitivity
]
```
So the max, will,
- first get the tuples where the `sp` is max, if unique returns it
- second, if multiple `sp` have same max value, get the tuple where `sn` is max

I'm a bit confused as to why this behaviour is incorrect. We want max possible specificity where `sensitivity >= min_sensitivity` right. If we do not mind sensitivity taking on any value that is greater than `min_sensitivity` then this is fine I guess.
Otherwise, we could also try something like this to be more clear,
```python
        # get indices where sensitivity is greater than min_sensitivity
        indices = sensitivity >= min_sensitivity

        # redefine specificity, sensitivity and threshold tensor based on indices
        specificity, sensitivity, thresholds = (
            specificity[indices], sensitivity[indices], thresholds[indices]
        )

        # get argmax
        idx = torch.argmax(specificity)
        
        # get max_spec and best_threshold
        max_spec, best_threshold = specificity[idx], thresholds[idx]
```

Comment (justusschock) [reply src/torchmetrics/functional/classification/specificity_at_sensitivity.py:53]: @shenoynikhil I prefer the second version actually, as this is more explicit in its behavior. That is, however, if it is not significantly slower.

Comment (shenoynikhil) [reply src/torchmetrics/functional/classification/specificity_at_sensitivity.py:53]: Wow, I'm a bit surprised that the second method is quite fast compared to the existing code. Maybe because everything is tensorized? 
This is the code I ran for timing the two methods,
```
import torch
import timeit
from typing import Tuple
from torch import Tensor
from src.torchmetrics.functional.classification.specificity_at_sensitivity import (
    _convert_fpr_to_specificity, 
    _binary_roc_compute,
    _binary_precision_recall_curve_format,
    _binary_precision_recall_curve_update
)

preds=torch.tensor([[0.2564, 0.4774, 0.8782, 0.7798, 0.7977, 0.7019, 0.4879, 0.1070, 0.3326,
         0.0214, 0.8027, 0.0783, 0.5409, 0.9140, 0.1960, 0.4214, 0.3306, 0.8577,
         0.8508, 0.4816, 0.4210, 0.6005, 0.4220, 0.3260, 0.1008, 0.4482, 0.6682,
         0.5491, 0.5096, 0.3329, 0.9925, 0.5386],
        [0.5321, 0.6358, 0.3810, 0.7685, 0.6603, 0.4452, 0.0417, 0.2789, 0.6787,
         0.8435, 0.4107, 0.6047, 0.8374, 0.6430, 0.8889, 0.7387, 0.1778, 0.3571,
         0.5001, 0.9463, 0.3610, 0.7706, 0.4914, 0.5994, 0.9341, 0.4619, 0.2995,
         0.4796, 0.8656, 0.9973, 0.9724, 0.6990],
        [0.6406, 0.6151, 0.2873, 0.9691, 0.5378, 0.4909, 0.1969, 0.2295, 0.7523,
         0.4090, 0.7889, 0.7036, 0.6282, 0.0759, 0.0450, 0.7602, 0.2913, 0.9583,
         0.8350, 0.4210, 0.3171, 0.0923, 0.7484, 0.0953, 0.5935, 0.8966, 0.0061,
         0.4136, 0.5178, 0.5872, 0.2090, 0.8769],
        [0.3409, 0.4287, 0.3359, 0.5169, 0.6149, 0.9185, 0.5165, 0.8586, 0.8909,
         0.4041, 0.9023, 0.8844, 0.6534, 0.9633, 0.1305, 0.1287, 0.9600, 0.9836,
         0.5763, 0.7460, 0.3046, 0.5831, 0.4937, 0.0988, 0.7049, 0.8960, 0.1797,
         0.0555, 0.7618, 0.3833, 0.2327, 0.4686]])
target=torch.tensor([[1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0,
         0, 1, 1, 1, 0, 1, 1, 0],
        [1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1,
         1, 0, 1, 0, 1, 1, 1, 1],
        [1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0,
         1, 0, 1, 0, 0, 0, 1, 1],
        [0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0,
         1, 0, 1, 1, 0, 1, 0, 1]])

preds, target, thresholds = _binary_precision_recall_curve_format(preds, target, thresholds=None, ignore_index=None)
state = _binary_precision_recall_curve_update(preds, target, thresholds)
fpr, sensitivity, thresholds = _binary_roc_compute(state, thresholds, pos_label = 1)
specificity = _convert_fpr_to_specificity(fpr)

# Method 1: Original
def _specificity_at_sensitivity_method_1(
    specificity: Tensor,
    sensitivity: Tensor,
    thresholds: Tensor,
    min_sensitivity: float,
) -> Tuple[Tensor, Tensor]:
    try:
        max_spec, _, best_threshold = max(
            [(sp, sn, thresh) for sp, sn, thresh in zip(specificity, sensitivity, thresholds) if sn >= min_sensitivity]
        )

    except ValueError:
        max_spec = torch.tensor(0.0, device=specificity.device, dtype=specificity.dtype)
        best_threshold = torch.tensor(1e6, device=thresholds.device, dtype=thresholds.dtype)

    return max_spec, best_threshold

# Method 2: More descriptive and using torch
def _specificity_at_sensitivity_method_2(
    specificity: Tensor,
    sensitivity: Tensor,
    thresholds: Tensor,
    min_sensitivity: float,
) -> Tuple[Tensor, Tensor]:
    try:
        
        # get indices where sensitivity is greater than min_sensitivity
        indices = sensitivity >= min_sensitivity

        # redefine specificity, sensitivity and threshold tensor based on indices
        specificity, sensitivity, thresholds = (
            specificity[indices], sensitivity[indices], thresholds[indices]
        )

        # get argmax
        idx = torch.argmax(specificity)
        
        # get max_spec and best_threshold
        max_spec, best_threshold = specificity[idx], thresholds[idx]

    except ValueError:
        max_spec = torch.tensor(0.0, device=specificity.device, dtype=specificity.dtype)
        best_threshold = torch.tensor(1e6, device=thresholds.device, dtype=thresholds.dtype)

    return max_spec, best_threshold

print ('Method: 1')
print (timeit.repeat('_specificity_at_sensitivity_method_1(specificity, sensitivity, thresholds, min_sensitivity=0.9)', globals=globals(), number=10000))

print ()
print ('Method: 2')
print (timeit.repeat('_specificity_at_sensitivity_method_2(specificity, sensitivity, thresholds, min_sensitivity=0.9)', globals=globals(), number=10000))

>> Method: 1
>> [8.020462333000001, 8.022529082999998, 8.109080291000001, 8.108721834, 8.068527583000005]
>> 
>> Method: 2
>> [0.28303899999999516, 0.2857496669999975, 0.2787325000000038, 0.2843657499999992, 0.2831222089999983]
```

Comment (shenoynikhil) [reply src/torchmetrics/functional/classification/specificity_at_sensitivity.py]: Have updated the code with the code I suggested. It's faster and more readable.

Comment (shenoynikhil) [reply src/torchmetrics/functional/classification/specificity_at_sensitivity.py]: I think tests are failing. Will try to fix them.

Comment (shenoynikhil) [reply src/torchmetrics/functional/classification/specificity_at_sensitivity.py]: @justusschock The (py)tests have been fixed. Feel free to look at the changes. There are some other tests that fail (probably outside the scope of this PR). Let me know your opinion.

Comment (justusschock) [reply src/torchmetrics/functional/classification/specificity_at_sensitivity.py]: @shenoynikhil looking good from my side. Just one more comment :)

File: src/torchmetrics/functional/classification/specificity_at_sensitivity.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+        # redefine specificity, sensitivity and threshold tensor based on indices
+        specificity, sensitivity, thresholds = (specificity[indices], sensitivity[indices], thresholds[indices])
+
+        # get argmax
+        idx = torch.argmax(specificity)
+
+        # get max_spec and best_threshold
+        max_spec, best_threshold = specificity[idx], thresholds[idx]
+
+    except:
```
Comment (justusschock) [inline src/torchmetrics/functional/classification/specificity_at_sensitivity.py:65]: can you specify the error-type here? We don't want to catch all of them, just the one that might rightfully occur here.

Comment (shenoynikhil) [reply src/torchmetrics/functional/classification/specificity_at_sensitivity.py:65]: I have an alternative to the error-type based solution, I checked locally and it passes all tests,

```python
    # get indices where sensitivity is greater than min_sensitivity
    indices = sensitivity >= min_sensitivity

    # if no indices are found, max_spec, best_threshold = 0.0, 1e6
    if not indices.any():
        max_spec = torch.tensor(0.0, device=specificity.device, dtype=specificity.dtype)
        best_threshold = torch.tensor(1e6, device=thresholds.device, dtype=thresholds.dtype)
    else:
        # redefine specificity, sensitivity and threshold tensor based on indices
        specificity, sensitivity, thresholds = specificity[indices], sensitivity[indices], thresholds[indices]

        # get argmax
        idx = torch.argmax(specificity)

        # get max_spec and best_threshold
        max_spec, best_threshold = specificity[idx], thresholds[idx]
```

The error comes up when there are no indices where `sensitivity >= min_sensitivity` and we try to do an `argmax` over an empty specificity list.

Comment (justusschock) [reply src/torchmetrics/functional/classification/specificity_at_sensitivity.py:65]: yes, please go ahead, push it and afterwards please run `pre-commit` and push these as well. Should be ready to go afterwards :)

Comment (shenoynikhil) [reply src/torchmetrics/functional/classification/specificity_at_sensitivity.py]: Done.

File: src/torchmetrics/classification/specificity_at_sensitivity.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+        >>> from torchmetrics.classification import BinarySpecificityAtSensitivity
+        >>> import torch
+        >>> preds = torch.tensor([0, 0.5, 0.4, 0.1])
+        >>> target = torch.tensor([0, 1, 1, 1])
+        >>> metric = BinarySpecificityAtSensitivity(min_sensitivity=0.5, thresholds=None)
+        >>> metric(preds, target)
+        (tensor(1.), tensor(0.4000))
+        >>> metric = BinarySpecificityAtSensitivity(min_sensitivity=0.5, thresholds=5)
+        >>> metric(preds, target)
+        (tensor(1.), tensor(0.2500))
```
Comment (Borda) [inline src/torchmetrics/classification/specificity_at_sensitivity.py:90]:
```suggestion
        >>> from torchmetrics.classification import BinarySpecificityAtSensitivity
        >>> from torch import tensor
        >>> preds = tensor([0, 0.5, 0.4, 0.1])
        >>> target = tensor([0, 1, 1, 1])
        >>> metric = BinarySpecificityAtSensitivity(min_sensitivity=0.5, thresholds=None)
        >>> metric(preds, target)
        (tensor(1.), tensor(0.4000))
        >>> metric = BinarySpecificityAtSensitivity(min_sensitivity=0.5, thresholds=5)
        >>> metric(preds, target)
        (tensor(1.), tensor(0.2500))
```

File: src/torchmetrics/classification/specificity_at_sensitivity.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+        ...                       [0.05, 0.75, 0.05, 0.05, 0.05],
+        ...                       [0.05, 0.05, 0.75, 0.05, 0.05],
+        ...                       [0.05, 0.05, 0.05, 0.75, 0.05]])
+        >>> target = torch.tensor([0, 1, 3, 2])
+        >>> metric = MulticlassSpecificityAtSensitivity(num_classes=5, min_sensitivity=0.5, thresholds=None)
+        >>> metric(preds, target)
+        (tensor([1., 1., 0., 0., 0.]), tensor([7.5000e-01, 7.5000e-01, 5.0000e-02, 5.0000e-02, 1.0000e+06]))
+        >>> metric = MulticlassSpecificityAtSensitivity(num_classes=5, min_sensitivity=0.5, thresholds=5)
+        >>> metric(preds, target)
+        (tensor([1., 1., 0., 0., 0.]), tensor([7.5000e-01, 7.5000e-01, 0.0000e+00, 0.0000e+00, 1.0000e+06]))
```
Comment (Borda) [inline src/torchmetrics/classification/specificity_at_sensitivity.py:178]:
```suggestion
        >>> from torch import tensor
        >>> preds = tensor([[0.75, 0.05, 0.05, 0.05, 0.05],
        ...                 [0.05, 0.75, 0.05, 0.05, 0.05],
        ...                 [0.05, 0.05, 0.75, 0.05, 0.05],
        ...                 [0.05, 0.05, 0.05, 0.75, 0.05]])
        >>> target = tensor([0, 1, 3, 2])
        >>> metric = MulticlassSpecificityAtSensitivity(num_classes=5, min_sensitivity=0.5, thresholds=None)
        >>> metric(preds, target)
        (tensor([1., 1., 0., 0., 0.]), tensor([7.5000e-01, 7.5000e-01, 5.0000e-02, 5.0000e-02, 1.0000e+06]))
        >>> metric = MulticlassSpecificityAtSensitivity(num_classes=5, min_sensitivity=0.5, thresholds=5)
        >>> metric(preds, target)
        (tensor([1., 1., 0., 0., 0.]), tensor([7.5000e-01, 7.5000e-01, 0.0000e+00, 0.0000e+00, 1.0000e+06]))
```

File: src/torchmetrics/classification/specificity_at_sensitivity.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+        >>> from torchmetrics.classification import MultilabelSpecificityAtSensitivity
+        >>> import torch
+        >>> preds = torch.tensor([[0.75, 0.05, 0.35],
+        ...                       [0.45, 0.75, 0.05],
+        ...                       [0.05, 0.55, 0.75],
+        ...                       [0.05, 0.65, 0.05]])
+        >>> target = torch.tensor([[1, 0, 1],
+        ...                        [0, 0, 0],
+        ...                        [0, 1, 1],
+        ...                        [1, 1, 1]])
```
Comment (Borda) [inline src/torchmetrics/classification/specificity_at_sensitivity.py:269]:
```suggestion
        >>> from torch import tensor
        >>> preds = tensor([[0.75, 0.05, 0.35],
        ...                 [0.45, 0.75, 0.05],
        ...                 [0.05, 0.55, 0.75],
        ...                 [0.05, 0.65, 0.05]])
        >>> target = tensor([[1, 0, 1],
        ...                  [0, 0, 0],
        ...                  [0, 1, 1],
        ...                  [1, 1, 1]])
```

**Review by Borda — APPROVED** [review]
